### ISSUE TITLE
[question] Finding conan package folder of dependency from conanfile.py

### ISSUE THREAD

**Issue by Donotalo** [conan-io/conan#15963]
### What is your question?

We're using conan v1.59 in our project (forced to as that's client's setup). It's a project in C, the build tool is CMake v3.22. This project has several dependencies. We're utilizing conanfile.py for building and packaging artifacts.

Let's say, ch-common is a requirement for this project. It's mentioned in conanfile.py as follows:
```
def requirements(self):
    ...
    self.requires("ch-common/0.0.10@x/y")
    ...
```

The development environment is Debian Linux. ch-common is downloaded and CMake can find it, code compiles properly. ch-common only has a header file cr_def.h.

The output of our project is a library. Internally the library uses an enum in cr_def.h. The users of our project (those who'll be calling various functions of the library) will also need the cr_def.h because in many APIs defined in the library the enum is used as parameter.

We have a package() in conanfile.py to package the artifacts:
```
def package(self):
    cmake = CMake(self)
    cmake.install()
    self.copy("*.h", dst="include/project/", src="include/public/", keep_path=False)
```
We'd like to copy cr_def.h in the package() like others.

How can cr_def.h be referred in the package() without hardcoded fixed path? We **don't want** to do something like:
```
self.copy("cr_def.h",
          dst="include/project/",
          src=os.path.expanduser('~') + "/.conan/data/ch-common/0.0.10/x/y/package/<hash>/include/",
          keep_path=False)
```
Is there any variable holding the package location (i.e., `os.path.expanduser('~') + "/.conan/data/ch-common/0.0.10/x/y/package/<hash>/`)? Is there any better way to copy cr_def.h?

We can't ask our library users to include ch-common in their project - that's a client requirement.

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @Donotalo 

Thanks for your question.

> How can cr_def.h be referred in the package() without hardcoded fixed path? We don't want to do something like:

I am afraid it is not fully clear what is happening. Definitely you shouldn't need to do anything like ``src=os.path.expanduser('~') + "/.conan/data/<full-cache-path>``, why would you do that?

The normal ``copy()`` should work, and the consumers should get it easily too, just the ``includedirs`` which defaults to ``["include"]`` already will allow to use the header in the package. Requirements in Conan are transitive too, not really necessary to re-package headers in consumers if this is what you are trying.

Could you please provide a more detailed description of the issue, the recipes and packages involved? I suspect that there might be more than one package recipe, but it is not clear. Is the issue that ``self.copy("*.h", dst="include/project/", src="include/public/", keep_path=False)`` is not packaging the header inside the package? Is the issue that the consumer of this package doesn't see the header even if it is inside the package?


Side note regarding:

>  self.copy("*.h", dst="include/project/", src="include/public/", keep_path=False)

Please try to move to ``copy(self, `` modern alternative, even if in Conan 1.X, to make sure the recipes are ready for Conan 2, which is now more than 1 year old.

**Comment by Donotalo** [comment]
Hi @memsharded,

Thank you for your response. As this is a client code, it's not possible to share. Before creating a bare minimum version of the problem, please allow me to further explain the issue. I'm new to conan, so creating a minimum version will take considerable amount of time. :)

Because of the following:
```
def requirements(self):
    ...
    self.requires("ch-common/0.0.10@x/y")
    ...
```
When I build the project, a header file called cr_defs.h is downloaded in `os.path.expanduser('~') + "/.conan/data/<full-cache-path>/include`. Because cr_defs.h is part of ch-common.

Take a look at the package() in this tutorial: https://docs.conan.io/2.1/tutorial/creating_packages/package_method.html

This is how header files are copied during packaging, from the link above:
```
copy(self, pattern="*.h", src=os.path.join(self.source_folder, "include"), dst=os.path.join(self.package_folder, "include"))
```

My problem is I don't know how to refer `os.path.expanduser('~') + "/.conan/data/<full-cache-path>/include` in the `src` parameter above.
```
copy(self, pattern="*.h", src=?????, dst=os.path.join(self.package_folder, "include"))
```

I'm trying to copy the cr_defs.h file, which is not part of my project, but comes from a requirement (ch-common). If there's another way please suggest.

In other words, my project depends on another project (ch-common) which contains a header file (cr_defs.h) that I need to package with my project.

**Comment by Donotalo** [comment]
I've found a way, but I couldn't find any official documentation so not sure if this is the correct way or not. In the `package()` method, I can access the include directory of `ch-common` as follows:
```
def package(self):
    ...
    self.copy("*.h", dst="include/project/",
        src=self.deps_cpp_info._dependencies['ch-common']._include_paths[0], keep_path=False)
    ...
```

**Comment by memsharded** [comment]
> My problem is I don't know how to refer os.path.expanduser('~') + "/.conan/data/<full-cache-path>/include in the src parameter above.

Please don't ever try to refer to internal cache paths that way, that is an implementation detail, and using it is undefined behavior. Paths should always be relative to the recipe, and using the different accessors: ``self.build_folder``, ``self.package_folder``, ``self.source_folder``, etc.

> self.deps_cpp_info

Please don't use this, this is quite legacy and has been superseded by ``self.dependencies["ch-common"].cpp_info.includedir``


If I understood correctly you are basically trying to "repackage" the headers inside ``ch-common`` inside a consumer package of ``ch-common``. This is what shouldn't be necessary, being ``ch-common`` a dependency, the consumer will already get access to those headers, without necessarily copying them, they are propagated transitively and automatically, without the consumer recipe having to make a copy of them. This is one of the issues that it is not very clear why you are doing that, it shouldn't be necessary in most cases.

**Comment by Donotalo** [comment]
> Please don't ever try to refer to internal cache paths that way, that is an implementation detail, and using it is undefined behavior. Paths should always be relative to the recipe, and using the different accessors: `self.build_folder`, `self.package_folder`, `self.source_folder`, etc.

No, we're not doing it. I mentioned it just to be clear what I'm trying to copy.

> 
> > self.deps_cpp_info
> 
> Please don't use this, this is quite legacy and has been superseded by `self.dependencies["ch-common"].cpp_info.includedir`
> 

`self.dependencies` doesn't exist. Remember, we're forced to use conan v1.59. Is it something available in v1.59?

> If I understood correctly you are basically trying to "repackage" the headers inside `ch-common` inside a consumer package of `ch-common`. This is what shouldn't be necessary, being `ch-common` a dependency, the consumer will already get access to those headers, without necessarily copying them, they are propagated transitively and automatically, without the consumer recipe having to make a copy of them. This is one of the issues that it is not very clear why you are doing that, it shouldn't be necessary in most cases.

I don't understand what you meant by "consumer package". The project, which is a library, I'm working on has a requirement of "ch-common". So the library can find the header files of "ch-common". However, the users of the library will also need the header files from "ch-common". These header files aren't automatically packaged, so we need to manually copy them.

**Comment by memsharded** [comment]
> No, we're not doing it. I mentioned it just to be clear what I'm trying to copy.

Great :)

> self.dependencies doesn't exist. Remember, we're forced to use conan v1.59. Is it something available in v1.59?

Yes: https://docs.conan.io/en/1.59/reference/conanfile/dependencies.html, since Conan 1.38, many years ago now.

> I don't understand what you meant by "consumer package". The project, which is a library, I'm working on has a requirement of "ch-common". So the library can find the header files of "ch-common". However, the users of the library will also need the header files from "ch-common". These header files aren't automatically packaged, so we need to manually copy them.

Lets say that the recipe with ``self.requires("ch-common/...")`` is being used to create the package ``mypkg/0.1``. When some other recipe does ``self.requires("mypkg/0.1")``, they will also automatically get ``ch-common`` and its headers, without having to do anything else, no need to do a copy of the headers inside ``mypkg/0.1``. Does this clarify it?

**Comment by Donotalo** [comment]
> > self.dependencies doesn't exist. Remember, we're forced to use conan v1.59. Is it something available in v1.59?
> 
> Yes: https://docs.conan.io/en/1.59/reference/conanfile/dependencies.html, since Conan 1.38, many years ago now.

Strange. When I try to access `self.dependencies` from `package()`, it throws the following error:

```
print(self.dependencies)
AttributeError: 'NoneType' object has no attribute 'dependencies'
```

But I can access other fields, like `self.deps_cpp_info._dependencies['ch-common']._include_paths[0]`. Not sure what's going on.

> Lets say that the recipe with `self.requires("ch-common/...")` is being used to create the package `mypkg/0.1`. When some other recipe does `self.requires("mypkg/0.1")`, they will also automatically get `ch-common` and its headers, without having to do anything else, no need to do a copy of the headers inside `mypkg/0.1`. Does this clarify it?

Yes. But how to make this happen? When I build the project, the header files from `ch-common` aren't automatically copied to `package` folder. The recipe has `self.requires("ch-common/...")` in the `requirements()` function.

**Comment by memsharded** [comment]
> Strange. When I try to access self.dependencies from package(), it throws the following error:

Definitely very strange. If you have a full recipe or a minimal repro case, we could give it a try. In code it should be there: https://github.com/conan-io/conan/blob/b6bcdc5f24bb9ab5e4ca8143a42637b1f010d471/conans/model/conan_file.py#L187, it is very unexpected

> Yes. But how to make this happen? When I build the project, the header files from ch-common aren't automatically copied to package folder. The recipe has self.requires("ch-common/...") in the requirements() function.

When you use some generator from Conan, if it is a transitive dependency, the ``includedirs`` information from that dependency is also generated for the consumer, so it can locate those headers too. In general, packages do not need to worry about doing anything with they dependencies to make them available to the final consumers, because this is automatic. If this is not happening, there might be some misconfiguration or issue with the generator and/or the build system scripts, but this would need something to reproduce too to be able to diagnose it.
